contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,541 @@
+      config.minimumSubscriptionBalance,
+      config.fallbackLinkPrice
+    );
+  }
+
+  // Want to ensure these arguments can fit inside of 2 words
+  // so in the worse case where the consuming contract has to read all of them
+  // from storage, it only has to read 2 words.
+  function requestRandomWords(
+    bytes32 keyHash,  // Corresponds to a particular offchain job which uses that key for the proofs
+    uint64  subId,
+    uint16  minimumRequestConfirmations,
+    uint32  callbackGasLimit,
+    uint32  numWords,  // Desired number of random words
+    uint32  consumerId // Index into consumers to avoid SLOADing all the consumers
+  )
+    external
+    returns (
+      uint256 requestId
+    )
+  {
+    // Input validation using the subscription storage.
+    if (s_subscriptions[subId].owner == address(0)) {
+      revert InvalidSubscription();
+    }
+    // We use this consumer index to ensure that the cost remains constant no matter
+    // the number of consumers.
+    // Its important to ensure that the consumer is in fact who they say they
+    // are, otherwise they could use someone else's subscription balance.
+    if (s_subscriptions[subId].consumers[consumerId] != msg.sender) {
+      revert InvalidConsumer(msg.sender);
+    }
+    // Input validation using the config storage word.
+    if (minimumRequestConfirmations < s_config.minimumRequestBlockConfirmations) {
+      revert RequestBlockConfsTooLow(minimumRequestConfirmations, s_config.minimumRequestBlockConfirmations);
+    }
+    if (s_subscriptions[subId].balance < s_config.minimumSubscriptionBalance) {
+      revert InsufficientBalance();
+    }
+    if (callbackGasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(callbackGasLimit, s_config.maxGasLimit);
+    }
+    // We could additionally check s_serviceAgreements[keyHash] != address(0)
+    // but that would require reading another word of storage. To save gas
+    // we leave that out.
+    uint256 nonce = s_nonces[keyHash][msg.sender] + 1;
+    uint256 preSeedAndRequestId = uint256(keccak256(abi.encode(keyHash, msg.sender, nonce)));
+
+    s_callbacks[preSeedAndRequestId] = keccak256(abi.encodePacked(
+        preSeedAndRequestId,
+        block.number,
+        subId,
+        callbackGasLimit,
+        numWords,
+        msg.sender));
+    emit RandomWordsRequested(keyHash, preSeedAndRequestId, subId, minimumRequestConfirmations, callbackGasLimit, numWords, msg.sender);
+    s_nonces[keyHash][msg.sender] = nonce;
+
+    return preSeedAndRequestId;
+  }
+
+  function getCallback(
+      uint256 requestId
+  )
+    external
+    view
+    returns (
+      bytes32
+    )
+  {
+    return s_callbacks[requestId];
+  }
+
+  // Offsets into fulfillRandomnessRequest's proof of various values
+  //
+  // Public key. Skips byte array's length prefix.
+  uint256 public constant PUBLIC_KEY_OFFSET = 0x20;
+  // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
```

Comment:
why not use this calculation for defining PRESEED_OFFSET and let the compiler eval it for us?